python/foglamp/plugins/south/http_south/http_south.py

```diff
@@ -10,21 +10,25 @@
 import os
 import ssl
 import logging
-
+from threading import Thread
 from aiohttp import web
 
 from foglamp.common import logger
 from foglamp.common.web import middleware
 from foglamp.plugins.common import utils
 from foglamp.services.south.ingest import Ingest
+import ingest
+
 
 __author__ = "Amarendra K Sinha"
 __copyright__ = "Copyright (c) 2017 Dianomic Systems"
 __license__ = "Apache 2.0"
 __version__ = "${VERSION}"
 
 _LOGGER = logger.setup(__name__, level=logging.INFO)
-
+c_callback = None
+c_ingest_ref = None
+loop = None
```

Comment:
Can we avoid to make these global? and pass these as an object's attributes to opaque handle?

Reply:
Two considerations:
- loop will remain essentially global as it shared between threads. 
- If we want callback and ingest_ref to be part of handle, that can be done by South C server itself when it calls plugin_init() with config argument in shim layer. This will also save us one method plugin_register_ingest() in shim layer. Otherwise how are we going to make plugin aware of callback and ingest_ref objects? Also, there is no use of keeping  plugin_register_ingest() also and then make callback and ingest_ref part of handle artificially within plugin code. But since this code architecture has been borrowed from C++, let us keep plugin code architecture consistent between Python async plugins and C++ async plugins.

Reply:
@praveen-garg
praveen-garg 2 days ago Member

Can we avoid to make these global? and pass these as an object's attributes to opaque handle?
@amarendra-dianomic
amarendra-dianomic 2 days ago Member

Two considerations:

    loop will remain essentially global as it shared between threads.
    If we want callback and ingest_ref to be part of handle, that can be done by South C server itself when it calls plugin_init() with config argument in shim layer. This will also save us one method plugin_register_ingest() in shim layer. Otherwise how are we going to make plugin aware of callback and ingest_ref objects? Also, there is no use of keeping plugin_register_ingest() also and then make callback and ingest_ref part of handle artificially within plugin code. But since this code architecture has been borrowed from C++, let us keep plugin code architecture consistent between Python async plugins and C++ async plugins.

@praveen-garg
praveen-garg 6 minutes ago Member

let's refactor it later.
Might be via shim layer or encapsulate c_callback, c_ingest_ref and c_ingest_ref and start start in thread with run target in some object and use that in handle.
@praveen-garg